cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java

```diff
@@ -259,61 +263,82 @@ private void increaseTopicReplicationFactor(ZkUtils zkUtils,
-          int cursor = cursors[rackCursor];
-          newAssignedReplica.add(brokersByRack.get(rack).get(cursor));
-          cursors[rackCursor] = (cursor + 1) % brokersByRack.get(rack).size();
+      if (pm.replicas().size() < replicationFactor) {
+        List<Object> newAssignedReplica = new ArrayList<>();
+        Set<String> currentOccupiedRack = new HashSet<>();
+        // Make sure the current replicas are in new replica list.
+        pm.replicas().forEach(node -> {
+          newAssignedReplica.add(node.id());
+          currentOccupiedRack.add(rackByBroker.get(node.id()));
+        });
+        // Add new replica to partition in rack-aware(if rack info is available), round-robin way.
+        while (newAssignedReplica.size() < replicationFactor) {
+          if (!currentOccupiedRack.contains(racks.get(rackCursor))) {
+            String rack = racks.get(rackCursor);
+            int cursor = cursors[rackCursor];
+            newAssignedReplica.add(brokersByRack.get(rack).get(cursor));
+            cursors[rackCursor] = (cursor + 1) % brokersByRack.get(rack).size();
+          }
+          rackCursor = (rackCursor + 1) % racks.size();
         }
-        rackCursor = (rackCursor + 1) % racks.size();
+        newReplicaAssignment.put(new TopicAndPartition(topic, pm.partition()),
+                                 JavaConverters.asScalaIteratorConverter(newAssignedReplica.iterator()).asScala().toSeq());
       }
-      newReplicaAssignment.put(pm.partition(), JavaConverters.asScalaIteratorConverter(newAssignedReplica.iterator()).asScala().toSeq());
     }
-    AdminUtils.createOrUpdateTopicPartitionAssignmentPathInZK(zkUtils, topic, newReplicaAssignment, props, true);
-    LOG.info("The replication factor of Kafka topic " + topic + " has increased to " + replicationFactor + ".");
+    if (newReplicaAssignment.nonEmpty()) {
+      zkUtils.updatePartitionReassignmentData(newReplicaAssignment);
+      LOG.info("The replication factor of Kafka topic " + topic + " has increased to " + replicationFactor + ".");
+    }
   }
 
-  private void ensureTopicCreated(ZkUtils zkUtils, Set<String> allTopics, String topic, long retentionMs, int replicationFactor, int partitionCount) {
+  /**
+   * Add new partitions to the Kafka topic.
+   *
+   * @param zkUtils ZkUtils class to use to increase replication factor.
+   * @param topic The topic to apply the change.
+   * @param topicMetadata Topic metadata stored in Zookeeper.
+   * @param partitionCount The target partition count of the topic.
+   */
+  private void maybeIncreaseTopicPartitionCount(ZkUtils zkUtils,
+                                                String topic,
+                                                MetadataResponse.TopicMetadata topicMetadata,
+                                                int partitionCount) {
+    if (partitionCount > topicMetadata.partitionMetadata().size()) {
+      if (!ensureTopicNotUnderPartitionReassignment(zkUtils, topic)) {
```

Comment:
Is checking for this topic sufficient or should we also check for (1) both sample store topics or (2) all topics?

Reply:
check for the topic is enough in this method me may only write to `brokers/topics/topic_name` znode to expand the AR list of this topic.